### PR TITLE
Add DefaultShardManager#setEventManagerProvider and JDA#getEventManager

### PR DESCRIPTION
**Pull request by maltesermailo**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines][contributing].

### Changes

- [X] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: #784

## Description
This pull request adds an option to add an EventManager Provider to the ShardManager so it does not register each shards listeners with the same EventManager.
Also it makes the method getEventManager in JDAImpl public to JDA

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManagerBuilder.java [hunk 1]
```diff
@@ -481,6 +484,31 @@ public DefaultShardManagerBuilder setEventManager(final IEventManager manager)
...
+     *         {@link net.dv8tion.jda.core.hooks.SubscribeEvent @SubscribeEvent} to mark the methods that listen for events.</li>
+     * </ul>
+     * <br>You can also create your own EventManager (See {@link net.dv8tion.jda.core.hooks.IEventManager}).
+     *
+     * @param  eventManagerProvider
+     *         A supplier for the new {@link net.dv8tion.jda.core.hooks.IEventManager} to use.
+     *
+     * @return The DefaultShardManagerBuilder instance. Useful for chaining.
+     */
+    public DefaultShardManagerBuilder setEventManagerProvider(final Supplier<IEventManager> eventManagerProvider) {
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManagerBuilder.java]: Braces on new lines

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManagerBuilder.java]: This should be `Supplier<? extends IEventManager>`

File: src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java [hunk 2]
```diff
@@ -637,6 +637,12 @@ public Presence getPresence()
     {
         return presence;
     }
+    
+    @Override
+    public IEventManager getEventManager()
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java]: Why was this moved?

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java [hunk 3]
```diff
@@ -647,7 +655,9 @@ protected JDAImpl buildInstance(final int shardId) throws LoginException, Interr
 
         jda.asBot().setShardManager(this);
 
-        if (this.eventManager != null)
+        if(this.eventManagerProvider != null)
+        	jda.setEventManager(this.eventManagerProvider.get());
+        else if (this.eventManager != null)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: There is no need to have both of these, you can simply make the setEventManager method create a provider for it.

**Comment by kantenkugel** [comment]
@maltesermailo can you explain why you marked the review from Minn as resolved? because you did not resolve the stuff he pointed out yet.

File: src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java (hunk 2 above)
Comment (maltesermailo) [reply src/main/java/net/dv8tion/jda/core/entities/impl/JDAImpl.java]: because it wasn't in the private area, i moved it to the functions where the override functions were

**Comment by maltesermailo** [comment]
> 
> 
> @maltesermailo can you explain why you marked the review from Minn as resolved? because you did not resolve the stuff he pointed out yet.

I thought it was a change as Git said he requested a change. After a question it was clear that this was review.

File: src/main/java/net/dv8tion/jda/core/JDA.java [hunk 4]
```diff
@@ -986,6 +986,12 @@ default Emote getEmoteById(long id)
     {
         return getEmoteCache().getElementsByName(name, ignoreCase);
     }
+    
+    /**
+     * Returns the instance of the EventManager used by this shard.
+     * @return The {@link net.dv8tion.jda.core.hooks.IEventManager} of this shard or null if none set.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/core/JDA.java]: This needs a newline between description and `@return` tag

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManagerBuilder.java [hunk 5]
```diff
@@ -478,8 +481,32 @@ public DefaultShardManagerBuilder setEventManager(final IEventManager manager)
...
+     *         {@link net.dv8tion.jda.core.hooks.SubscribeEvent @SubscribeEvent} to mark the methods that listen for events.</li>
+     * </ul>
+     * <br>You can also create your own EventManager (See {@link net.dv8tion.jda.core.hooks.IEventManager}).
+     *
+     * @param  eventManagerProvider
+     *         A supplier for the new {@link net.dv8tion.jda.core.hooks.IEventManager} to use.
+     *
+     * @return The DefaultShardManagerBuilder instance. Useful for chaining.
+     */
+    public DefaultShardManagerBuilder setEventManagerProvider(final IntFunction<? extends IEventManager> eventManagerProvider) {
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManagerBuilder.java]: Braces on new line

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java [hunk 6]
```diff
@@ -647,9 +649,9 @@ protected JDAImpl buildInstance(final int shardId) throws LoginException, Interr
 
         jda.asBot().setShardManager(this);
 
-        if (this.eventManager != null)
-            jda.setEventManager(this.eventManager);
-
+        if(this.eventManagerProvider != null)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: Can you add the space back that you removed here between `if` and the parenthesis

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java [hunk 7]
```diff
@@ -253,6 +254,7 @@
      *         The token
      * @param  eventManager
      *         The event manager
+     * @param eventManagerProvider
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: Missing space and description

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java (hunk 6 above)
Comment (kantenkugel) [reply src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: Also tab -> space in the line below

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java (hunk 7 above)
Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: Also you should probably remove the other event manager parameter since its gone now

File: src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java (hunk 6 above)
Comment (maltesermailo) [reply src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java]: fuck eclipse

**Comment by kantenkugel** [comment]
Should we deprecate the old `setEventManager()`? Since i feel like its rather harmful and just using `setEventManagerProvider()` is an easy fix and not for some reason not apply-able to code that previously used `setEventManager()`
